docs/orchestration.asciidoc

```diff
@@ -0,0 +1,106 @@
+- To replace dedicated master and dedicated data nodes by nodes having both master and data roles, replace the 2 existing NodeSets by a single one with a different name and the corresponding Elasticsearch configuration settings.
+- To upgrade Elasticsearch from version `7.2.0` to `7.3.0`, simply change the value in the `version` field.
+
+No downtime is expected when NodeSet changes are orchestrated by ECK. It ensures that:
+
+- data on a node that is being removed is completely migrated to other nodes before actually removing said node
+- Elasticsearch orchestration settings such as `discovery.seed_hosts`, `cluster.initial_master_nodes`, `discovery.zen.minimum_master_nodes`, `_cluster/voting_config_exclusions` are adjusted as necessary for any cluster topology change
+- rolling upgrades are performed safely, reusing PersistentVolumes of the upgraded Elasticsearch nodes
+
+[id="{p}-statefulsets"]
+==== StatefulSets orchestration
+
+Behind the scenes, ECK translates each NodeSet specified in the Elasticsearch resource into a link:https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/[StatefulSet in Kubernetes]. The StatefulSet specification is based on the NodeSet specification:
+
+* `count` corresponds to the number of replicas in the StatefulSet, each replica leading to the creation of a Pod, which corresponds to a single Elasticsearch node
+* `podTemplate` can be used to specify custom settings for the Elasticsearch Pod, overriding the default ones set by ECK on the generated StatefulSet specification
+* the StatefulSet name is built from the Elasticsearch resource name and the NodeSet name. Each Pod will be assigned the StatefulSet name suffixed by an ordinal. The corresponding Elasticsearch node has the same name as the Pod.
+
+The actual Pod creation is handled by the StatefulSet controller in Kubernetes. ECK relies on the link:https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/#on-delete[OnDelete StatefulSet update strategy] since it needs full control over when and how Pods get upgraded to a new revision.
+
+Whenever a Pod is removed and recreated (maybe with a newer revision), the StatefulSet controller ensures PersistentVolumes that were attached to the original Pod are attached again to the newly created Pod.
+
+[id="{p}-upgrade-patterns"]
+==== Cluster upgrade patterns
+
+Depending on how the NodeSets are updated, ECK handles the Kubernetes resources reconciliation in various ways.
+
+* When a new NodeSet is added to the Elasticsearch resource, ECK creates the corresponding StatefulSet. It also sets up link:https://kubernetes.io/docs/concepts/configuration/secret/[Secrets] and link:https://kubernetes.io/docs/tasks/configure-pod-container/configure-pod-configmap/[ConfigMaps] to hold the TLS certificates and Elasticsearch configuration files.
```

Comment:
I think that for advanced users or for support team it would be interesting to say that these actions are done in this exact order.

Reply:
I think that can also be confusing due to the nature of the reconciliation loop. It may execute step 1 (scale up), then return an error, then skip step 1 because there's nothing else to do and go straight to step 2 (downscale) and return. Then because of the change budget get back to step 1 where we can add more nodes, etc.
It feels hard to explain correctly here?

Reply:
Yes may be that this document is not the right place. I think it is still a valuable information when it's time to understand what the operator is doing and predict what it will do.